src/EventStore.Core/Services/Storage/EpochManager/EpochManager.cs

```diff
@@ -177,33 +198,30 @@ public class EpochManager : IEpochManager {
 				if (sysRec.SystemRecordType != SystemRecordType.Epoch)
 					return false;
 
-				var epoch = sysRec.GetEpochRecord();
+				epoch = sysRec.GetEpochRecord();
 				return epoch.EpochNumber == epochNumber && epoch.EpochId == epochId;
 			} finally {
 				_readers.Return(reader);
 			}
 		}
 
 		// This method should be called from single thread.
-		public void WriteNewEpoch() {
-			// Set epoch checkpoint to -1, so if we crash after new epoch record was written, 
-			// but epoch checkpoint wasn't updated, on restart we don't miss the latest epoch.
-			// So on node start, if there is no epoch checkpoint or it contains negative position, 
-			// we do sequential scan from the end of TF to find the latest epoch record.
-			//NOTE AN: It seems we don't need to pessimistically set epoch checkpoint to -1, because
-			//NOTE AN: if crash occurs in the middle of writing epoch or updating epoch checkpoint,
-			//NOTE AN: then on restart we'll start from chaser checkpoint (which is not updated yet)
-			//NOTE AN: and process all records till the writer checkpoint, so all epochs will be processed 
-			//NOTE AN: and epoch checkpoint will ultimately contain correct last epoch position. This process
-			//NOTE AN: is similar to index rebuild process.
-			//_checkpoint.Write(-1);
-			//_checkpoint.Flush();
-
+		public void WriteNewEpoch(int epochNumber) {
 			// Now we write epoch record (with possible retry, if we are at the end of chunk) 
 			// and update EpochManager's state, by adjusting cache of records, epoch count and un-caching 
 			// excessive record, if present.
 			// If we are writing the very first epoch, last position will be -1.
-			var epoch = WriteEpochRecordWithRetry(_lastEpochNumber + 1, Guid.NewGuid(), _lastEpochPosition, _instanceId);
+			if (epochNumber < 0) {
+				throw new ArgumentException($"Cannot add write Epoch with a negative Epoch Number {epochNumber}.",nameof(epochNumber));
+			}
+			if (epochNumber < LastEpochNumber) {
+				throw new ArgumentException($"Cannot add write Epoch with a Epoch Number {epochNumber} lower than Last Epoch Number {LastEpochNumber}.",nameof(epochNumber));
+			}
+			if (epochNumber <= LastEpochNumber) {
+				throw new Exception($"Concurrency failure, Last Epoch Number is {LastEpochNumber} cannot add write new Epoch with a lower Epoch Number {epochNumber}.");
+			}
```

Comment:
@condron  The purpose of this `if` statement is unclear, could you please clarify? If assuming that there could be any concurrency issue it could still happen after this if statement and before we call `WriteEpochRecordWithRetry`.

Reply:
These if statements are just to give better exception messages, I've reworded the section.